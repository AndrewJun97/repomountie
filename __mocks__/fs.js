// __mocks__/fs.js
'use strict';

const fs = jest.requireActual('fs');

function access(path, flag, cb) {
  cb(undefined);
}

function readFile(path, options, cb) {
  cb(undefined, Buffer.from('Hello World', 'utf8'));
  // cb(undefined, 'Hello World');
}

fs.readFile = readFile;
fs.access = access;

module.exports = fs;
