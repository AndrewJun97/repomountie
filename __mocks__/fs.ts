// __mocks__/fs.js
'use strict';

const fs = require.requireActual('fs');

function access(path, flag, cb) {
  if (path === 'no-file-access') {
    return cb(new Error('No access to this file - mock'));
  }

  return cb(undefined);
}

function readFile(path, options, cb) {
  if (path === 'no-file') {
    return cb(new Error('No such file - mock'), undefined);
  }
  const text = 'Hello World';
  return cb(undefined, typeof options === 'undefined' ? Buffer.from(text, 'utf8') : text);
}

fs.readFile = readFile;
fs.access = access;

module.exports = fs;
