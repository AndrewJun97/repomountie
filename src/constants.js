"use strict";
//
// Copyright Â© 2018, 2019, 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//
exports.__esModule = true;
exports.SCHEDULER_DELAY = 24 * 60 * 60 * 1000; // one day
exports.COMMENT_TRIGGER_WORD = 'help';
exports.GITHUB_ID = 'repo-mountie';
exports.REPO_CONFIG_FILE = 'rmconfig.json';
exports.HELP_DESK = {
    SUPPORT_USERS: ['jleach']
};
exports.COMMIT_FILE_NAMES = {
    COMPLIANCE: 'COMPLIANCE.yaml',
    LICENSE: 'LICENSE'
};
exports.COMMIT_MESSAGES = {
    ADD_COMPLIANCE: 'Add compliance audit file',
    ADD_LICENSE: 'Add Apache License 2.0',
    UPDATE_COMPLIANCE: 'Updating compliance audit file'
};
exports.PR_TITLES = {
    ADD_COMPLIANCE: 'Add missing compliance audit file',
    ADD_LICENSE: 'Add missing license'
};
exports.BRANCHES = {
    ADD_COMPLIANCE: 'repo-mountie/add-compliance',
    ADD_LICENSE: 'repo-mountie/add-license'
};
exports.TEMPLATES = {
    COMPLIANCE: 'templates/COMPLIANCE.yaml',
    CONDUCT: 'templates/CODE_OF_CONDUCT.md',
    CONTRIBUTE: 'templates/CONTRIBUTING.md',
    LICENSE: 'templates/LICENSE',
    README: 'templates/README.md'
};
exports.VALID_LICENSES = {
    APACHE: 'apache-2.0'
};
exports.TEXT_FILES = {
    HOWTO_PR: 'templates/howto_pull_request.md',
    STALE_COMMENT: 'templates/stale_issue_comment.md',
    WHY_COMPLY: 'templates/why-comply.md',
    WHY_LICENSE: 'templates/why-license.md'
};
exports.COMMANDS = {
    IGNORE: '/bot-ignore-length'
};
exports.REGEXP = {
    command: '@repo-mountie\\s+',
    compliance: '@repo-mountie\\s+update-(pia|stra)\\s+(in-progress|completed|TBD|exempt)',
    help: '@repo-mountie\\s+help'
};
exports.ACCESS_CONTROL = {
    allowedInstallations: ['bcgov', 'fullboar'],
    allowedSsoClients: ['reggie-api']
};
