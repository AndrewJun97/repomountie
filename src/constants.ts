//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

export const SCHEDULER_DELAY: number = 24 * 60 * 60 * 1000; // one day

export const HELP_DESK = {
  LICENSE_SUPPORT_USERS: ['jleach'],
};

export const PR_TITLES = {
  ADD_LICENSE: 'Add missing license',
};

export const BRANCHES = {
  ADD_LICENSE: 'fix/add-license',
};

export const TEMPLATES = {
  CONDUCT: 'templates/CODE_OF_CONDUCT.md',
  CONTRIBUTE: 'templates/CONTRIBUTING.md',
  LICENSE: 'templates/LICENSE',
  README: 'templates/README.md',
};

export const VALID_LICENSES = {
  APACHE: 'apache-2.0',
};

export const TEXT_FILES = {
  WHY_LICENSE: 'templates/why-license.md',
};

export const ALLOWED_REPOS = [
  'repo-mountie',
  'devhub-signing-api',
  'devhub-signing-web',
  'devhub-signing-agent',
  'range-api',
  'range-web',
];
