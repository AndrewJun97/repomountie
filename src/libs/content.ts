//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/nodejs-common-utils';
import { Context } from 'probot';
import { BRANCHES, TEMPLATES } from '../constants';
import { loadTemplate } from './utils';

/**
 * Add a license file via PR to a given repo
 *
 * @param {Context} context The repo context
 * @returns
 */
export const addLicense = async (context: Context) => {
  const commitMessage: string = 'Add Apache License 2.0';
  const body: string = `
    Repos in our organization need to be licensed under the Apache License 2.0.
    To help you get up to spec I've added one for you as part of this PR;
    please merge it when you can. If you have an exception please add comments to
    this PR and close it without merging it.`;

  try {
    // If we don't have a master we won't have know where to merge the PR
    const master = await context.github.gitdata.getReference(
      context.repo({
        ref: 'heads/master',
      })
    );

    if (!master) {
      return;
    }

    // Create a branch to commit to commit the license file
    await context.github.gitdata.createReference(
      context.repo({
        ref: BRANCHES.LICENSE,
        sha: master.data.object.sha,
      })
    );

    const data = await loadTemplate(TEMPLATES.LICENSE);
    if (!data) {
      return;
    }

    // Add the file to the new branch
    await context.github.repos.createFile(
      context.repo({
        branch: BRANCHES.LICENSE,
        content: Buffer.from(data).toString('base64'),
        message: commitMessage,
        path: 'LICENSE',
      })
    );

    // Create a PR to merge the licence ref into master
    await context.github.pullRequests.create(
      context.repo({
        base: 'master',
        body,
        head: BRANCHES.LICENSE,
        maintainer_can_modify: true, // maintainers cat edit your this PR
        title: 'Add missing license',
      })
    );
  } catch (err) {
    logger.log(err.message);

    throw err;
  }
};
