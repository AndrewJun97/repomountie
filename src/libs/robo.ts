//
// Repo Mountie
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2019-12-04.
//

import { logger } from '@bcgov/common-nodejs-utils';
import yaml from 'js-yaml';
import { Context } from 'probot';
import { BRANCHES, COMMIT_FILE_NAMES, COMMIT_MESSAGES, HELP_DESK, PR_TITLES, REGEXP } from '../constants';
import { assignUsersToIssue, fetchContentsForFile } from './ghutils';

/**
 * Determine if help desk support is required
 * @param {Context} context The event context context
 * @returns True if support is required, false otherwise.
 */
export const helpDeskSupportRequired = (payload: any) => {
    // These are the accepted commands. They are case insensitive,
    // and require a leading `@re` to be accepted.
    // @${WHOAMI} help

    const re = new RegExp(REGEXP.help, 'gi');
    const comment = payload.comment.body;
    const isAssigned = payload.issue.assignees.some((e) =>
        HELP_DESK.SUPPORT_USERS.includes(e.login)
    );

    if (!isAssigned && re.test(comment)) {
        return true;
    }

    return false;
};

export const applyComplianceCommands = (comment: string, doc: any): any => {
    let result: RegExpExecArray | null;
    const re = new RegExp(REGEXP.compliance, 'gi');

    // tslint:disable-next-line:no-conditional-assignment
    while ((result = re.exec(comment)) !== null) {
        // sample result
        // [ '/update-pia completed',
        //   'pia',
        //   'completed',
        //   index: 0,
        //   input: '/update-pia completed',
        //   groups: undefined ]
        const items = doc.spec.filter((s) => s.name === result![1].toUpperCase());
        if (items.length === 0) {
            continue;
        }

        const item = items.pop();
        item.status = result![2];
        // Date.now() used to simplify mock in unit tests.
        item['last-updated'] = new Date(Date.now()).toISOString();
    }

    return doc;
};

export const handleLicenseCommands = async (context: Context) => {

    try {
        // For we just assign help desk willy-nilly, going forward we should
        // better identify the issue and assign users with surgical precision.
        if (helpDeskSupportRequired(context.payload)) {
            await assignUsersToIssue(context, HELP_DESK.SUPPORT_USERS);
        }
    } catch (err) {
        const message = 'Unable to process license commands';
        logger.error(`${message}, error = ${err.message}`);

        throw err;
    }
};

export const handleComplianceCommands = async (context: Context) => {
    // These are the accepted commands. They are case insensitive,
    // and require a leading `@re` to be accepted.
    // @${WHOAMI} update-pia STATUS
    // @${WHOAMI} update-stra STATUS
    const re = new RegExp(REGEXP.compliance, 'gi');
    const comment = context.payload.comment.body;

    try {
        // For we just assign help desk willy-nilly, going forward we should
        // better identify the issue and assign users with surgical precision.
        if (helpDeskSupportRequired(context.payload)) {
            await assignUsersToIssue(context, HELP_DESK.SUPPORT_USERS);
        }

        // check for appropriate bot commands
        if (!re.test(comment)) {
            return;
        }

        // Fetch the file from the repo in case any updates have
        // been made to it.
        const data = await fetchContentsForFile(context,
            COMMIT_FILE_NAMES.COMPLIANCE, BRANCHES.ADD_COMPLIANCE);

        if (!data) {
            logger.info(`Unable to fetch ${COMMIT_FILE_NAMES.COMPLIANCE} in ref ${BRANCHES.ADD_COMPLIANCE}`);
            return;
        }

        let doc = yaml.safeLoad(Buffer.from(data.content, 'base64').toString());
        doc = applyComplianceCommands(comment, doc);

        await context.github.repos.createOrUpdateFile(
            context.repo({
                branch: BRANCHES.ADD_COMPLIANCE,
                content: Buffer.from(yaml.safeDump(doc)).toString('base64'),
                message: COMMIT_MESSAGES.UPDATE_COMPLIANCE,
                path: COMMIT_FILE_NAMES.COMPLIANCE,
                sha: data.sha,
            })
        );
    } catch (err) {
        const message = 'Unable to process compliance commands';
        logger.error(`${message}, error = ${err.message}`);

        throw err;
    }
};

export const handleBotCommand = async (context: Context) => {

    if (context.isBot) {
        return; // not interested
    }

    try {
        switch (context.payload.issue.title) {
            case PR_TITLES.ADD_COMPLIANCE:
                await handleComplianceCommands(context);
                break;
            case PR_TITLES.ADD_LICENSE:
                await handleLicenseCommands(context);
                break;
            default:
                break;
        }
    } catch (err) {
        const message = 'Unable to process commands';
        logger.error(`${message}, error = ${err.message}`);

        throw err;
    }
};
