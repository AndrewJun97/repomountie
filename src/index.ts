//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/nodejs-common-utils';
import createScheduler from '@bcgov/probot-scheduler';
import { Application, Context } from 'probot';
import { SCHEDULER_DELAY } from './constants';
import { created } from './libs/issue';
import { validatePullRequestIfRequired } from './libs/pullrequest';
import { addLicenseIfRequired } from './libs/repository';

process.env.TZ = 'UTC';

export = (app: Application) => {
  logger.info('Robot Loaded!!!');

  const scheduler = createScheduler(app, {
    delay: false, // !!process.env.DISABLE_DELAY, // delay is enabled on first run
    interval: SCHEDULER_DELAY,
  });

  app.on('pull_request.opened', pullRequestOpened);
  app.on('issue_comment.created', issueCommentCreated);
  app.on('schedule.repository', repositoryScheduled);
  app.on('repository.deleted', repositoryDelete);

  async function pullRequestOpened(context: Context) {
    logger.info(
      `Processing PR ${context.payload.pull_request.number} for repo ${
        context.payload.repository.name
      }`
    );

    try {
      // This can throw a `TypeError` during testing.
      if (context.isBot) {
        // Don't act crazy.
        return;
      }
    } catch (err) {
      logger.info('Unable to determine if the sender is a bot');
    }

    await validatePullRequestIfRequired(context);
  }

  async function issueCommentCreated(context: Context) {
    logger.info(`Processing issue ${context.payload.issue.id}`);

    try {
      // This can throw a `TypeError` during testing.
      if (context.isBot) {
        // Don't act crazy.
        return;
      }
    } catch (err) {
      logger.info('Unable to determine if the sender is a bot');
    }

    try {
      await created(context);
    } catch (err) {
      logger.error(`Unable to process issue ${context.payload.issue.id}`);
    }
  }

  async function repositoryDelete(context: Context) {
    scheduler.stop(context.payload.repository);
  }

  async function repositoryScheduled(context: Context) {
    logger.info(`Processing ${context.payload.repository.name}`);

    try {
      if (context.payload.repository.archived) {
        logger.warn(`The repo ${context.payload.repository.name} is archived. Skipping.`);
        return;
      }

      await addLicenseIfRequired(context, scheduler);
    } catch (err) {
      logger.error(`Unable to add license to ${context.payload.repository.name}`);
    }
  }
};
