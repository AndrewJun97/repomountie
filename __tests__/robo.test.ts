//
// Repo Mountie
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2019-12-04.
//

import fs from 'fs';
import path from 'path';
import { GITHUB_ID } from '../src/constants';
import { handleComplianceCommands, helpDeskSupportRequired } from '../src/libs/robo';

const p0 = path.join(__dirname, 'fixtures/issue_comment-event.json');
const context = JSON.parse(fs.readFileSync(p0, 'utf8'));

describe('Bot command processing', () => {

    beforeEach(() => {
        context.payload.issue.user.login = `${GITHUB_ID}[bot]`;
        context.payload.comment.body = 'I\'m a teapot';
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    // TODO:(jl) test handleBotCommand

    it('Help requests issues are processed', async () => {
        context.payload.comment.body = 'help';
        const result = helpDeskSupportRequired(context.payload);

        expect(result).toBeTruthy();
    });

    it('Non-help comments are ignored', async () => {
        const result = helpDeskSupportRequired(context.payload);

        expect(result).toBeFalsy();
    });

    it('Blarb', async () => {
        handleComplianceCommands('body');
    });
});
