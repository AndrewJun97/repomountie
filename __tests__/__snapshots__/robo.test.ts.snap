// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bot command processing Invalid compliance commands are ignored 1`] = `
Object {
  "description": "This document is used to track a projects PIA and STRA
compliance.
",
  "name": "compliance",
  "spec": Array [
    Object {
      "last-updated": "2019-11-22T00:03:52.138Z",
      "name": "PIA",
      "status": "TBD",
    },
    Object {
      "last-updated": "2019-11-22T00:03:52.138Z",
      "name": "STRA",
      "status": "TBD",
    },
  ],
}
`;

exports[`Bot command processing Missing compliance commands are ignored 1`] = `
Object {
  "description": "This document is used to track a projects PIA and STRA
compliance.
",
  "name": "compliance",
  "spec": Array [
    Object {
      "last-updated": "2019-11-22T00:03:52.138Z",
      "name": "PIA",
      "status": "TBD",
    },
    Object {
      "last-updated": "2019-11-22T00:03:52.138Z",
      "name": "STRA",
      "status": "TBD",
    },
  ],
}
`;

exports[`Bot command processing Valid compliance commands are processed 1`] = `
Object {
  "description": "This document is used to track a projects PIA and STRA
compliance.
",
  "name": "compliance",
  "spec": Array [
    Object {
      "last-updated": "2019-12-11T18:58:32.480Z",
      "name": "PIA",
      "status": "completed",
    },
    Object {
      "last-updated": "2019-12-11T18:58:32.480Z",
      "name": "STRA",
      "status": "completed",
    },
  ],
}
`;
